File: using-HTTP.py
#!/usr/bin/env python3
"""
PTZOptics HTTP-CGI API Controller Example

This example demonstrates how to control PTZOptics cameras using HTTP-CGI commands.
Commands are sent automatically every 5 seconds in a cycle.

Usage:
    python using-HTTP.py

Requirements:
    - Python 3.6+
    - requests library (pip install requests)
    - PTZOptics camera on network
    - Camera IP address or hostname configured below
"""

import requests
import time
import threading
import signal
import sys
from urllib.parse import urljoin
from requests.auth import HTTPDigestAuth

# Configuration Constants
CAMERA_IP = "192.168.15.164"
CAMERA_USERNAME = "admin"
CAMERA_PASSWORD = "admin1"

class PTZOpticsHTTPController:
    def __init__(self, host="ptzoptics.local", username="admin", password="admin"):
        self.host = host
        self.base_url = f"http://{host}"
        self.username = username
        self.password = password
        self.running = False
        self.command_index = 0
        self.timer_thread = None
        self.session = requests.Session()
        
        # Set up digest authentication if credentials provided
        if username and password:
            self.session.auth = HTTPDigestAuth(username, password)
        
        # HTTP-CGI Command List - Add or modify commands here
        self.cgi_commands = [
            # Pan Left (medium speed: pan=12, tilt=10)
            {
                "url": "/cgi-bin/ptzctrl.cgi?ptzcmd&left&12&10",
                "params": {},
                "method": "GET"
            },
            
            # Stop Pan/Tilt
            {
                "url": "/cgi-bin/ptzctrl.cgi?ptzcmd&ptzstop&0&0",
                "params": {},
                "method": "GET"
            },
            # Pan Left (medium speed: pan=12, tilt=10)
            {
                "url": "/cgi-bin/ptzctrl.cgi?ptzcmd&right&12&10",
                "params": {},
                "method": "GET"
            },
        ]
        
        # Command descriptions for logging
        self.command_descriptions = [
            "Pan Left (Speed 12/10)",
            "Stop Pan/Tilt"
            "Pan Right (Speed 12/10)",
        ]
    
    def test_connection(self):
        """Test connection to the PTZ camera"""
        try:
            # Test with a simple ptzctrl stop command
            test_url = urljoin(self.base_url, "/cgi-bin/ptzctrl.cgi?ptzcmd&ptzstop&0&0")
            response = self.session.get(test_url, timeout=10)
            
            if response.status_code == 200:
                print(f"✓ Connected to PTZ camera at {self.host}")
                return True
            elif response.status_code == 401:
                print(f"❌ Authentication failed - check username/password")
                return False
            else:
                print(f"❌ Connection failed - HTTP {response.status_code}")
                return False
                
        except requests.RequestException as e:
            print(f"❌ Connection error: {e}")
            return False
    
    def send_command(self, command):
        """Send an HTTP-CGI command to the camera"""
        try:
            # Build the full URL
            url = urljoin(self.base_url, command["url"])
            
            # Display the URL being sent
            print(f"   URL: {command['url']}")
            
            # Send the request
            if command["method"].upper() == "POST":
                response = self.session.post(url, data=command["params"], timeout=10)
            else:
                response = self.session.get(url, params=command["params"] if command["params"] else None, timeout=10)
            
            # Check response
            if response.status_code == 200:
                print(f"   ✅ Command successful (HTTP {response.status_code})")
                if response.text.strip():
                    # Show first line of response if available
                    first_line = response.text.strip().split('\n')[0][:100]
                    print(f"   Response: {first_line}...")
                return True
            else:
                print(f"   ❌ Command failed (HTTP {response.status_code})")
                if response.text:
                    print(f"   Error: {response.text[:200]}...")
                return False
                
        except requests.RequestException as e:
            print(f"   ❌ Request error: {e}")
            return False
    
    def send_next_command(self):
        """Send the next command in the cycle"""
        if not self.cgi_commands:
            print("⚠️ No commands in command list")
            return
        
        command = self.cgi_commands[self.command_index]
        description = (self.command_descriptions[self.command_index] 
                      if self.command_index < len(self.command_descriptions) 
                      else f"Command {self.command_index + 1}")
        
        print(f"\n📤 Sending: {description}")
        self.send_command(command)
        
        # Move to next command (cycle through the list)
        self.command_index = (self.command_index + 1) % len(self.cgi_commands)
    
    def command_timer(self):
        """Timer function that sends commands every 5 seconds"""
        while self.running:
            if self.running:  # Double-check we're still running
                self.send_next_command()
            time.sleep(5)  # 5 second interval
    
    def start(self):
        """Start the controller"""
        if self.test_connection():
            self.running = True
            self.timer_thread = threading.Thread(target=self.command_timer, daemon=True)
            self.timer_thread.start()
            print("🕒 Started command timer (5 second intervals)")
            return True
        return False
    
    def stop(self):
        """Stop the controller"""
        self.running = False
        if self.timer_thread:
            self.timer_thread.join(timeout=1)
        print("⏹️ Controller stopped")
    
    def send_single_command(self, index):
        """Send a single command by index"""
        if 0 <= index < len(self.cgi_commands):
            command = self.cgi_commands[index]
            description = (self.command_descriptions[index] 
                          if index < len(self.command_descriptions) 
                          else f"Command {index + 1}")
            
            print(f"\n📤 Manual send: {description}")
            self.send_command(command)
        else:
            print(f"❌ Invalid command index: {index}")

def signal_handler(sig, frame):
    """Handle Ctrl+C gracefully"""
    print("\n\n🛑 Stopping PTZ Controller...")
    controller.stop()
    sys.exit(0)

def main():
    global controller
    
    # Create controller instance
    controller = PTZOpticsHTTPController(
        host=CAMERA_IP,
        username=CAMERA_USERNAME,
        password=CAMERA_PASSWORD
    )
    
    # Set up signal handler for graceful shutdown
    signal.signal(signal.SIGINT, signal_handler)
    
    print("PTZOptics HTTP-CGI Controller Example")
    print("=" * 40)
    print(f"Connecting to camera at: {controller.host}")
    print("Commands will be sent every 5 seconds")
    print("Press Ctrl+C to stop\n")
    
    # Start the controller
    if controller.start():
        try:
            # Keep the program running
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            signal_handler(None, None)
    else:
        print("Failed to start controller")
        sys.exit(1)

if __name__ == "__main__":
    main()

"""
HOW TO ADD MORE COMMANDS:

1. Add command dictionaries to the 'cgi_commands' list
2. Add corresponding descriptions to 'command_descriptions' list

Example commands from the PTZOptics HTTP-CGI documentation:

Pan/Tilt Commands:
- Pan Right:     {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "right", "12": "", "10": ""}, "method": "GET"}
- Tilt Up:       {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "up", "12": "", "10": ""}, "method": "GET"}
- Tilt Down:     {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "down", "12": "", "10": ""}, "method": "GET"}
- Up-Left:       {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "leftup", "12": "", "10": ""}, "method": "GET"}
- Up-Right:      {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "rightup", "12": "", "10": ""}, "method": "GET"}
- Down-Left:     {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "leftdown", "12": "", "10": ""}, "method": "GET"}
- Down-Right:    {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "rightdown", "12": "", "10": ""}, "method": "GET"}

Zoom Commands:
- Zoom In:       {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "zoomin"}, "method": "GET"}
- Zoom Out:      {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "zoomout"}, "method": "GET"}
- Zoom Stop:     {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "zoomstop", "0": ""}, "method": "GET"}

Focus Commands:
- Focus In:      {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "focusin"}, "method": "GET"}
- Focus Out:     {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "focusout"}, "method": "GET"}
- Focus Stop:    {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "focusstop", "0": ""}, "method": "GET"}

Preset Commands:
- Call Home:     {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "home"}, "method": "GET"}
- Set Preset 1:  {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "posset", "1": ""}, "method": "GET"}
- Call Preset 1: {"url": "/cgi-bin/ptzctrl.cgi", "params": {"ptzcmd": "poscall", "1": ""}, "method": "GET"}

Speed Parameters (for pan/tilt):
- Pan speed: 1 (slowest) to 24 (fastest)
- Tilt speed: 1 (slowest) to 20 (fastest)
- Medium speeds: Pan=12, Tilt=10

Installation:
- Install requests: pip install requests
- Update camera credentials in main() function
- Modify the 5-second interval in command_timer() if needed

Note: Some commands may require POST method with Content-Length header.
See the HTTP-CGI documentation for POST request formatting requirements.
"""

File: node/using-HTTP.js
#!/usr/bin/env node
/**
 * PTZOptics HTTP-CGI API Controller Example
 * 
 * This example demonstrates how to control PTZOptics cameras using HTTP-CGI commands.
 * Implements HTTP Digest Authentication (SHA-256) natively using Node.js built-in modules.
 * 
 * Usage:
 *    node using-HTTP.js
 */

const http = require('http');
const crypto = require('crypto');

// Configuration Constants
const CAMERA_HOST = "192.168.1.100";
const CAMERA_USERNAME = "admin";
const CAMERA_PASSWORD = "admin";

class PTZOpticsHTTPController {
    constructor(host = "ptzoptics.local", username = "admin", password = "admin") {
        this.host = host;
        this.username = username;
        this.password = password;
        this.running = false;
        this.commandIndex = 0;
        this.timerInterval = null;
        
        // HTTP-CGI Command List
        this.cgiCommands = [
            "/cgi-bin/ptzctrl.cgi?ptzcmd&left&12&10",
            "/cgi-bin/ptzctrl.cgi?ptzcmd&ptzstop&0&0", 
            "/cgi-bin/ptzctrl.cgi?ptzcmd&right&12&10",
        ];
        
        // Command descriptions
        this.commandDescriptions = [
            "Pan Left (Speed 12/10)",
            "Stop Pan/Tilt",
            "Pan Right (Speed 12/10)",
        ];
    }
    
    makeRequest(path) {
        return new Promise((resolve, reject) => {
            const options = {
                hostname: this.host,
                port: 80,
                path: path,
                method: 'GET',
                headers: {
                    'Accept': '*/*'
                }
            };
            
            const req = http.request(options, (res) => {
                let data = '';
                
                res.on('data', chunk => {
                    data += chunk;
                });
                
                res.on('end', () => {
                    if (res.statusCode === 401 && res.headers['www-authenticate']) {
                        // Need to handle digest auth
                        const authHeader = res.headers['www-authenticate'];
                        const digestResponse = this.createDigestResponse(authHeader, path);
                        
                        // Make authenticated request
                        const authOptions = {
                            ...options,
                            headers: {
                                ...options.headers,
                                'Authorization': digestResponse
                            }
                        };
                        
                        const authReq = http.request(authOptions, (authRes) => {
                            let authData = '';
                            authRes.on('data', chunk => {
                                authData += chunk;
                            });
                            authRes.on('end', () => {
                                resolve({ 
                                    statusCode: authRes.statusCode, 
                                    data: authData,
                                    headers: authRes.headers 
                                });
                            });
                        });
                        
                        authReq.on('error', reject);
                        authReq.end();
                        
                    } else {
                        resolve({ 
                            statusCode: res.statusCode, 
                            data: data,
                            headers: res.headers 
                        });
                    }
                });
            });
            
            req.on('error', reject);
            req.end();
        });
    }
    
    createDigestResponse(authHeader, uri) {
        // Parse WWW-Authenticate header
        const params = {};
        authHeader.replace('Digest ', '').split(',').forEach(param => {
            const [key, value] = param.trim().split('=');
            params[key] = value ? value.replace(/"/g, '') : '';
        });
        
        const realm = params.realm || '';
        const nonce = params.nonce || '';
        const qop = params.qop || '';
        const algorithm = params.algorithm || 'MD5';
        
        // Generate client nonce
        const cnonce = crypto.randomBytes(8).toString('hex');
        const nc = '00000001';
        
        // Calculate digest
        let ha1, ha2, response;
        
        if (algorithm === 'SHA-256') {
            const hash = crypto.createHash('sha256');
            hash.update(`${this.username}:${realm}:${this.password}`);
            ha1 = hash.digest('hex');
            
            const hash2 = crypto.createHash('sha256');
            hash2.update(`GET:${uri}`);
            ha2 = hash2.digest('hex');
            
            const hash3 = crypto.createHash('sha256');
            if (qop === 'auth') {
                hash3.update(`${ha1}:${nonce}:${nc}:${cnonce}:${qop}:${ha2}`);
            } else {
                hash3.update(`${ha1}:${nonce}:${ha2}`);
            }
            response = hash3.digest('hex');
        } else {
            // MD5
            const hash = crypto.createHash('md5');
            hash.update(`${this.username}:${realm}:${this.password}`);
            ha1 = hash.digest('hex');
            
            const hash2 = crypto.createHash('md5');
            hash2.update(`GET:${uri}`);
            ha2 = hash2.digest('hex');
            
            const hash3 = crypto.createHash('md5');
            if (qop === 'auth') {
                hash3.update(`${ha1}:${nonce}:${nc}:${cnonce}:${qop}:${ha2}`);
            } else {
                hash3.update(`${ha1}:${nonce}:${ha2}`);
            }
            response = hash3.digest('hex');
        }
        
        // Build Authorization header
        let authStr = `Digest username="${this.username}", realm="${realm}", nonce="${nonce}", uri="${uri}"`;
        if (qop) {
            authStr += `, cnonce="${cnonce}", nc=${nc}, qop=${qop}`;
        }
        authStr += `, response="${response}", algorithm=${algorithm}`;
        
        return authStr;
    }
    
    async testConnection() {
        try {
            console.log(`Testing connection to: http://${this.host}/cgi-bin/ptzctrl.cgi?ptzcmd&ptzstop&0&0`);
            const result = await this.makeRequest('/cgi-bin/ptzctrl.cgi?ptzcmd&ptzstop&0&0');
            
            console.log(`Response status: ${result.statusCode}`);
            if (result.statusCode === 200) {
                console.log(`✓ Connected to PTZ camera at ${this.host}`);
                return true;
            } else {
                console.log(`❌ Connection failed - HTTP ${result.statusCode}`);
                console.log(`Response: ${result.data.substring(0, 200)}`);
                return false;
            }
        } catch (error) {
            console.log(`❌ Connection error: ${error.message}`);
            return false;
        }
    }
    
    async sendCommand(path) {
        try {
            console.log(`   URL: ${path}`);
            const result = await this.makeRequest(path);
            
            if (result.statusCode === 200) {
                console.log(`   ✅ Command successful (HTTP ${result.statusCode})`);
                if (result.data.trim()) {
                    const firstLine = result.data.trim().split('\n')[0].substring(0, 100);
                    console.log(`   Response: ${firstLine}`);
                }
                return true;
            } else {
                console.log(`   ❌ Command failed (HTTP ${result.statusCode})`);
                return false;
            }
        } catch (error) {
            console.log(`   ❌ Request error: ${error.message}`);
            return false;
        }
    }
    
    async sendNextCommand() {
        if (!this.cgiCommands || this.cgiCommands.length === 0) {
            console.log("⚠️ No commands in command list");
            return;
        }
        
        const command = this.cgiCommands[this.commandIndex];
        const description = this.commandDescriptions[this.commandIndex] || `Command ${this.commandIndex + 1}`;
        
        console.log(`\n📤 Sending: ${description}`);
        await this.sendCommand(command);
        
        this.commandIndex = (this.commandIndex + 1) % this.cgiCommands.length;
    }
    
    async start() {
        if (await this.testConnection()) {
            this.running = true;
            
            await this.sendNextCommand();
            
            this.timerInterval = setInterval(() => {
                if (this.running) {
                    this.sendNextCommand();
                }
            }, 5000);
            
            console.log("🕒 Started command timer (5 second intervals)");
            return true;
        }
        return false;
    }
    
    stop() {
        this.running = false;
        if (this.timerInterval) {
            clearInterval(this.timerInterval);
            this.timerInterval = null;
        }
        console.log("⏹️ Controller stopped");
    }
}

let controller;

function signalHandler() {
    console.log("\n\n🛑 Stopping PTZ Controller...");
    if (controller) {
        controller.stop();
    }
    process.exit(0);
}

async function main() {
    controller = new PTZOpticsHTTPController(CAMERA_HOST, CAMERA_USERNAME, CAMERA_PASSWORD);
    
    process.on('SIGINT', signalHandler);
    process.on('SIGTERM', signalHandler);
    
    console.log("PTZOptics HTTP-CGI Controller Example");
    console.log("=".repeat(40));
    console.log(`Connecting to camera at: ${controller.host}`);
    console.log("Commands will be sent every 5 seconds");
    console.log("Press Ctrl+C to stop\n");
    
    if (await controller.start()) {
        // Keep running
    } else {
        console.log("Failed to start controller");
        process.exit(1);
    }
}

main().catch((error) => {
    console.error("Fatal error:", error);
    process.exit(1);
});
